updated release workflow for V2

ivansenic opened the pull request:
**What this PR does**:

Main goal was to include `apis/` projects in the release by:
 * publishing docker images
 * bumping versions in the PR

However, I ended-up with a complete refactor of a release workflow.. Here is the overview on the jobs and updates made:

1. Create release
    * still creates release in the same way
    * however, `uploadUrl` is shared to other jobs as the output, no need for upload/download stuff
2. Build coordinator
    * still main task to upload `stargate-libs.zip` to the created release
    * now correctly depends on the `create-release` job and it's output
    * updated actions versions to latest (`actions/checkout@v3`, `actions/setup-java@v3`)
    * uses maven wrapper instead of maven
    * extra memory for maven for faster builds `MAVEN_OPTS: '-Xmx4g'`
3. Publish to OSSRH
    * part of the previous job `publish`, but just handling publishing the coordinator jars to the OSSRH
    * updated actions versions, use wrapper, etc
4. Publish docker images
    * part of the previous job `publish`, but just handling publishing the docker images
    * added support for publishing the `/api` as well, done with installing two java versions
    * same updates as everywhere else
5. Create PR for version update
    * part of the previous job `publish`, but just handling new PR for the new version
    * added support for updating the versions of `/api` as well
    * same updates as everywhere else

In addition I defined a manual trigger with the possibility to skip the publishing of the artifacts and docker images.

**Which issue(s) this PR fixes**:
Internal issue.

**Checklist**
- [x] Manual trigger inputs with no push

ivansenic commented on .github/workflows/release-v2.yml:
```diff
@@ -6,9 +6,14 @@ on:
 jobs:
+
+  # creates a new release if it's not existing
+  # outputs the upload URL in the releaseUploadUrl output var
```
wrong name of the var

ivansenic replied:
Done.

ivansenic commented:
Should this be on the `master`, so that I can actually test stuff?

mpenick commented on .github/workflows/release-v2.yml:
```diff
@@ -128,45 +149,190 @@ jobs:
-      - name: Update version numbers for next release
+
+      - name: Build and push (apis, Amazon ECR)
+        if: ${{ inputs.publish != false }}
```
`${{ inputs.publish == true }}` seems easier to understand compared to `${{ inputs.publish != false }}`. ` `${{ inputs.publish }}` not work, or does it need `!=` (or `==`) for coercion?

mpenick replied:
ah, nm, I understand, this work well for testing. :) Because absence is true.

ivansenic replied:
But how would this work when it's not a manual trigger and there are no inputs?

mpenick replied:
I didn't see your response. I understand, the absence is true to allow for testing.

ivansenic commented on .github/workflows/release-v2.yml:
```diff
@@ -63,48 +75,57 @@ jobs:
           ARTIFACTORY_PASSWORD: ${{ secrets.ARTIFACTORY_PASSWORD }}
           OSSRH_USERNAME: ${{ secrets.OSSRH_USERNAME }}
           OSSRH_TOKEN: ${{ secrets.OSSRH_TOKEN }}
+          MAVEN_OPTS: '-Xmx4g'
```
GH runner has 7GB of memory available, in my tests I figured-out that coordinator builds are like 2-3 times faster if I give more memory to the maven.. Seems that OSGi manifest creation is some kind of bottleneck there..

tatu-at-datastax replied:
Interesting that manifest creation is so slow. But I guess it depends on CLASSPATH size, and we have so many dependencies...

ivansenic commented on .github/workflows/release-v2.yml:
```diff
@@ -128,45 +149,190 @@ jobs:
-          mvn versions:set -DremoveSnapshot versions:commit && \
-          mvn -B -Dgpg.passphrase=${{ secrets.OSSRH_GPG_SECRET_KEY_PASSWORD }} clean deploy -DskipTests -P deploy,dse
+          cd apis/
+          JAVA_HOME=$JAVA_17 ./mvnw versions:set -DremoveSnapshot versions:commit
```
Yea basically I don't need it as I have afterwards:

```
JAVA_HOME=$JAVA_17 ./mvnw clean package -DskipTests -Dquarkus.container-image.build=true -Dquarkus.container-image.push=true -Dquarkus.container-image.tag=${{ steps.vars.outputs.tag }}
```

Which does clean package, build of docker and push in one command..

tatu-at-datastax replied:
So without specifying new version, does this just increase the patch number by one?

ivansenic replied:
Not even sure if it's needed to remove the `-SNAPSHOT` here, I am anyway tagging with the custom tag value?

tatu-at-datastax commented on .github/workflows/release-v2.yml:
```diff
@@ -20,31 +33,30 @@ jobs:
         with:
-          java-version: 1.8
+          distribution: 'temurin'
+          java-version: '8'
```
We can use either `8` for latest, or specify specific version like `8.0.332`. No strong opinion on how much to pin, although I did have issues with `ccm` failing on specific new version a while ago. But I guess we can pin to specific version if (and only if) JDK-related problems found.

tatu-at-datastax commented on .github/workflows/release-v2.yml:
```diff
@@ -63,48 +75,57 @@ jobs:
         with:
           ref: "v2.0.0"
+
+      # TODO this below is some magic I don't get, why not using ${{ github.ref }} as in create-release
```
Would it be easy to test to see if that'd work? The whole "echo" construct is needed for some reason by Github actions to set output variable for other steps to see I think. Sort of idiom although quite incomprehensible to me.

tatu-at-datastax commented on .github/workflows/release-v2.yml:
```diff
@@ -128,45 +149,190 @@ jobs:
+          cache: maven
+      - run: echo "JAVA_8=$JAVA_HOME" >> $GITHUB_ENV
+
+      - name: Set up JDK 8
```
Minor typo, should be 17.

ivansenic replied:
Done.

jeffreyscarpenter commented on .github/workflows/release-v2.yml:
```diff
@@ -128,45 +149,190 @@ jobs:
+    runs-on: ubuntu-latest
+    steps:
+
+      # TODO same here, why do we need the ref?
```
> Only a single commit is fetched by default, for the ref/SHA that triggered the workflow. 
source: https://github.com/actions/checkout

Tatu is right, this is brittle. Let's omit these `with` overrides.

ivansenic replied:
Actually I realized this is major bug at the moment. Consider a following situation:

* I tag locally a commit that should be released
* Anybody merges a PR that pushes another commit on a master
* I now push that local tag to the remote
* The release is not checking out the commit that is tagged by the release, but the one from the PR that's being merged..

:exploding_head:

jeffreyscarpenter reviewed (approved):
Lots of great improvements here, many of them would apply to the v1 release workflow as well.

ivansenic commented:
Added few further improvements:

* when `input.publish` is `false` I am still building the images, but not pushing them
* checkout the reference, as suggested by @tatu-at-datastax and @jeffreyscarpenter 
* added debug output for the tag resolving as it seems not needed

ivansenic commented:
Pushed additional updates:
* fix `MAVEN_OPTS` usage..
* introduce `inputs.tag` for specification of the release tag value when running from a manual trigger
* removed strange tag resolving in the publish jobs

ivansenic commented:
`default-push-trigger-1 (stargateio) ` failing due to the #2001. Merging anyway.
